hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightEndpoint.java

```diff
@@ -674,6 +685,624 @@ public Response getDeletedDirectorySummary() {
+      }
+
+      // If no keys were found, return a response indicating that no keys matched
+      if (!keysFound) {
+        return ReconResponseUtils.noMatchedKeysResponse(paramInfo.getStartPrefix());
+      }
+
+      // Set the aggregated totals in the response
+      listKeysResponse.setReplicatedDataSize(replicatedTotal);
+      listKeysResponse.setUnReplicatedDataSize(unreplicatedTotal);
+
+      return Response.ok(listKeysResponse).build();
+    } catch (IOException e) {
+      return ReconResponseUtils.createInternalServerErrorResponse(
+          "Error listing keys from OM DB: " + e.getMessage());
+    } catch (RuntimeException e) {
+      return ReconResponseUtils.createInternalServerErrorResponse(
+          "Unexpected runtime error while searching keys in OM DB: " + e.getMessage());
+    } catch (Exception e) {
+      return ReconResponseUtils.createInternalServerErrorResponse(
+          "Error listing keys from OM DB: " + e.getMessage());
+    }
+  }
+
+  public Map<String, OmKeyInfo> searchKeysInFSO(ParamInfo paramInfo)
+      throws IOException {
+    int originalLimit = paramInfo.getLimit();
+    Map<String, OmKeyInfo> matchedKeys = new LinkedHashMap<>();
+    // Convert the search prefix to an object path for FSO buckets
+    String startPrefixObjectPath = convertStartPrefixPathToObjectIdPath(paramInfo.getStartPrefix());
+    String[] names = parseRequestPath(startPrefixObjectPath);
+    Table<String, OmKeyInfo> fileTable =
+        omMetadataManager.getKeyTable(BucketLayout.FILE_SYSTEM_OPTIMIZED);
+
+    // If names.length > 2, then the search prefix is at the volume or bucket level hence
+    // no need to find parent or extract id's or find subpaths as the fileTable is
+    // suitable for volume and bucket level search
+    if (names.length > 2) {
+      // Fetch the parent ID to search for
+      long parentId = Long.parseLong(names[names.length - 1]);
+
+      // Fetch the nameSpaceSummary for the parent ID
+      NSSummary parentSummary =
+          reconNamespaceSummaryManager.getNSSummary(parentId);
+      if (parentSummary == null) {
+        return matchedKeys;
+      }
+      List<String> subPaths = new ArrayList<>();
+      // Add the initial search prefix object path because it can have both files and subdirectories with files.
+      subPaths.add(startPrefixObjectPath);
+
+      // Recursively gather all subpaths
+      gatherSubPaths(parentId, subPaths, Long.parseLong(names[0]), Long.parseLong(names[1]));
+      // Iterate over the subpaths and retrieve the files
+      for (String subPath : subPaths) {
+        paramInfo.setStartPrefix(subPath);
+        matchedKeys.putAll(
+            retrieveKeysFromTable(fileTable, paramInfo));
+        paramInfo.setLimit(originalLimit - matchedKeys.size());
+        if (matchedKeys.size() >= originalLimit) {
+          break;
+        }
+      }
+      return matchedKeys;
+    }
+
+    paramInfo.setStartPrefix(startPrefixObjectPath);
```

Comment:
Do this scenario will hit? at volume level, at request entry itself, check is there to have atleast volume and bucket.

Reply:
Yes, it will hit , if startPrefix path is till bucket. Pls check and run `testListKeysFSOBucket`